### PR TITLE
Implement limits per operation

### PR DESCRIPTION
**Pull request by Grenadingue**
Hi!
I though implementing this feature would be useful for one of my projects, so I'm sharing it here.
It's worth mentioning I've not yet extensively used it.
Would you mind sharing your thoughts and feedback about all this?

### CODE REVIEW

**Comment by devkral** [comment]
sry for the late answer: what are you trying to solve? There are some code cleanups but also some dangerous stuff like snakecase:
what is it for? It seems like you use the snakecase option (but why introducing a dependency for that? Regex should work too)

**Comment by Grenadingue** [comment]
> sry for the late answer

The same goes for me ^^

---

> what are you trying to solve?

I found this project to be useful for what it was intended, as stated in `master`'s
readme. However, I found some usage limitations to emerge from the
fact that the configurable limits are only global (project's wide). What if for
example, I would like my project's queries and mutations to globally have a fixed depth
limit of N, but for one special query I would like it to be N-1, or N+1, or anything else?
Therefore being able to override the global configuration per operation (per query or mutation)
made sense to me. Does that makes sense to you?

---

> some dangerous stuff like snakecase: what is it for?

About snakecase, before talking about it, I think it's better to understand "where" is the
case conversion occuring in the current PR's code.

`base.ProtectorBackend._current_operation_limits` method performs interospection on Python's
`Query`s and `Mutation`s objects, with the goal of finding optional limits per operation or
falling back to the default project's wide limits defined by `GRAPHENE_PROTECTOR_*` settings.

This procedure is achieved first by finding the name of the current GraphQL operation, so either
a query or a mutation name, and from this name, the method looks into `Query`s and `Mutation`s objects
to find attributes which have a name with the following format `f'limit_{operation_name}'`.
If this attribute exists, then its internal values, if present, will override the default settings for the current operation.

For example, if we take the following "users" query:

```python 3
class UserNode(DjangoObjectType):
   ...

class Query(graphene.ObjectType):
   users = DjangoFilterConnectionField(UserNode) # here
```

Then an adequate GraphQL query matching the previously defined `Query` would look like:

```graphql
query {
  users { // here
     ...
  }
}
```

And so, the "limit" attribute would be placed like this:

```python 3
...

class Query(graphene.ObjectType):
   users = DjangoFilterConnectionField(UserNode)
   limit_users = Limits(depth=42) # like this
```

The above procedure seems to work fine until we start using GraphQL operations with a name composed
of multiple words, like "create user". In this scenario, naming case conventions for the GraphQL operation and Python's
`Query`s and `Mutation`s objects' attributes stop matching. With my (brand new) current understanding,
usually GraphQL operations case convention is "camelCase", while Python's objects attributes case
convention is "snake_case". Without any case conversion, `base.ProtectorBackend._current_operation_limits`
fails to find `f'limit_{operation_name}'` attributes when their `{operation_name}` is composed of
multiple words. Here because `limit_createUser" does not exist. Therefore, in order to fix this, I introduced this case
conversion step, from "createUser" to "create_user".

To illustrate my example, lets consider the mutation request "create user":

```graphql
mutation {
  createUser(...) { // here
     ...
  }
}
```

And the associated definition of the operation and its limits:

```python 3
class CreateUser(graphene.Mutation):
   ...

class Mutation(graphene.AbstractType):
   create_user = CreateUser.Field()
   limit_create_user = Limits(depth=42) // here
   ...
```

We can see here that the case convention differs between `createUser` and `limit_create_user`.

With all this in mind, I see now that the current code does not answer the following questions:
- should we assume `Query` and `Mutation` objects attributes to always be named by following the snake case convention, in order to find each `f'limit_{operation_name}'` attribute?
  - should this be instead a configurable value in django settings like `GRAPHENE_PROTECTOR_ATTRIBUTE_CASE`?

---

> It seems like you use the snakecase option (but why introducing a dependency for that? Regex should work too)

About the stringcase library introduction as a dependency, honestly that an excellent question!

When I asked it to myself while fixing the case conversion issue, I was torn by these questions and arguments:
- should I use a dependency?
  - in order not to reinvent the wheel
  - and possibly to enjoy future maintenance of this dependency by others
- should I define a little set of tools by myself
  - in order to fully control what I'm doing
  - in the most efficient way as I can
  - but possibly reinventing the wheel
  - and eventually pushing my future myself to maintain this tool, while someone else could have done it for me

My thoughts were a lot guided by [this answer's comments on stackoverflow](https://stackoverflow.com/a/17328907).

There is visibly no clear side to stand on, so I favored the idea of not reinventing the wheel, of code factorization,
which seems to be a personal preference.

You seem to favor the opposite side, could you feed me with your arguments and preferences about this?

**Comment by devkral** [comment]
thank you, the idea is very good.

I would do it differently: why not expanding a resolver and intercepting special attributes from fields. Therefore we need no special object members.
I think of something like
field = Field(objtype)
field.complexity = 100
this would allow decorators or

field = Field(objtype, complexity=100)

we shouldn't pollute the object space as it opens security issues (how do we handle clashes?)

**Comment by Grenadingue** [comment]
Sorry for the long latency.

I see what you mean, I understand the downsides you described and find your suggestion practically very neat!

Could you expand your thoughts about how you would implement the resolver and what is a resolver here?

Also, I guess the graphene-protector's `Field` class you were describing should in a way or another possess the properties of others graphene `Field` class. Right? As far as I dug, it seems that there are not one but many `Field` class implementations (Pure Graphene's one [[1]](https://github.com/graphql-python/graphene/blob/c08379ed85b2759de32777eb5dd3dca143a6d69f/graphene/relay/connection.py#L179) [[2]](https://github.com/graphql-python/graphene/blob/c08379ed85b2759de32777eb5dd3dca143a6d69f/graphene/relay/connection.py#L119), Django's one [[1]](https://github.com/graphql-python/graphene-django/blob/623d0f219ebeaf2b11de4d7f79d84da8508197c8/graphene_django/fields.py#L74) [[2]](https://github.com/graphql-python/graphene-django/blob/623d0f219ebeaf2b11de4d7f79d84da8508197c8/graphene_django/filter/fields.py#L30), MongoDB [[1]](https://github.com/graphql-python/graphene-mongo/blob/3ef74bc81533fc6f3a2de9e17e5a0aa97fde1eb2/graphene_mongo/fields.py#L35), SQLAlchemy [[1]](https://github.com/graphql-python/graphene-sqlalchemy/blob/20ecaeadf2144b88555a3daf1a04e31b7f2ff95a/graphene_sqlalchemy/fields.py#L17) [[2]](https://github.com/graphql-python/graphene-sqlalchemy/blob/20ecaeadf2144b88555a3daf1a04e31b7f2ff95a/graphene_sqlalchemy/fields.py#L90) and Google AppEngine [[1]](https://github.com/graphql-python/graphene-gae/blob/d56dec0cfb31131d26bcca1431db3844c1dd2c07/graphene_gae/ndb/fields.py#L98)), all of them inheriting from the graphene's `ConnectionField` class.

Here is a class diagram I made in order to get some more sense of it.

class diagram with all Fields

Should we create something like a `ProtectedField` class and then create a new counterpart class for each `Field` class for the 3 levels at the bottom, extending the content of each `Field` class by inheritance? Thus giving birth to classes like `ProtectedDjangoConnectionField`, `ProtectedDjangoFilterConnectionField`, etc. Again, here is a diagram to get some more sense of it. For the sake of readability, it describes only what would happen to Graphene's `ConnectionField` and Django's `Field`s.

class diagram with protected Fields

I am no Python expert, so maybe there is a more Python-ish way to implement this? How would you implement the `ProtectedField` class(es) you referred to as `Field` in your example?

And also, how would this allow the use of decorators?

File: graphene_protector/base.py [hunk 1]
```diff
@@ -1,9 +1,48 @@
...
+
+    def __init__(self, *args, **kwargs):
+        # For every self attribute, get same attribute's name from kwargs, and save it if it exists
+        # Alternatively, use the already set value
+        # ie. if depth_limit=1 has been given as a kwarg, it will be saved as self.depth_limit
+        # else, self.depth_limit will be used as default value
+        for attribute, value in getmembers(self):
+            if not callable(attribute) and not attribute.startswith('_'):
+                self.__dict__[attribute] = kwargs.get(attribute, value)
+
```
Comment (devkral) [inline graphene_protector/base.py:24]: dataclasses is what you seeking:
https://docs.python.org/3/library/dataclasses.html

**Comment by devkral** [comment]
good point. To solve this problem I would suggest a per object helper dict

File: graphene_protector/base.py (hunk 1 above)
Comment (devkral) [reply graphene_protector/base.py:24]: do you still need python 3.6 support?

File: graphene_protector/base.py [hunk 2]
```diff
@@ -90,12 +122,46 @@ def document_from_string(self, schema, document_string):
             if not isinstance(definition, OperationDefinition):
                 continue
 
+            limits = self._current_operation_limits(schema, definition)
+
```
Comment (devkral) [inline graphene_protector/base.py:126]: a drawback: limits are not inheritated

**Comment by devkral** [comment]
today I had some time:
https://github.com/devkral/graphene-protector/tree/feature/per-field-limits
my idea: having a dict with all the limit information and create a decorator to fill the dict

WARNING: it is not ready yet, decorator is missing and needs to be tested

**Comment by devkral** [comment]
the branch I created is now functional. Can you try it?
I still need to write documentation.

**Comment by devkral** [comment]
For info: I inject a custom attribute into the fields, this seems to work reliable.

**Comment by Grenadingue** [comment]
Sure ! I will take a look by the end of the week

**Comment by devkral** [comment]
I fixed the issue, plus I fixed some security problems.
The library can behave a bit different.

**Comment by Grenadingue** [comment]
So I gave a try to the updated `master` branch, here is some feedback

- how to allow graphene's relay nodes `edges` to function while the `ProtectorBackend` is in use? how to set custom limits to them?
  - `AttributeError: type object 'UserNodeConnection' has no attribute 'edges'`
- how to allow schema introspection to function while the `ProtectorBackend` is in use? how to set custom limits to it?
  - `AttributeError: type object 'Query' has no attribute '__schema'`
- [`base.check_resource_usage`'s `limit.selections` condition](https://github.com/devkral/graphene-protector/blob/df1900d3aad4582c0dfe6805ce227d451d543731/graphene_protector/base.py#L120) did not exactly work for specific `Limits`
  - it seemed to me that the global `Limits` were used one `level` too deep, instead of the specific ones
  - I worked around the issue by renaming `sub_limits` to `limits`, so `limits` will be overriden, but I'm not sure that is the right solution
- the current implementation forced me to convert my `Query` & `Mutation` attributes with composed names from snake_case to camelCase
  - I believe this should be up to the programmer to choose which string case suits best his needs
- some more logging verbosity would be enjoyable, especially for error cases and requests rejection

Besides that, interesting implementation!

**Comment by devkral** [comment]
the camel case problems should be fixed.
to the other issues: the code is weird and needs definitely more documentation but have a look in tests. If the limits work correctly everything should be fine

**Comment by devkral** [comment]
I didn't tested it with relay. So there could be some issues

**Pull request closed without merging**